crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -0,0 +1,878 @@
+use crate::{
+    archetype::{Archetype, ArchetypeId, Archetypes},
+    bundle::{Bundle, BundleInfo, DynamicBundle},
+    component::{Component, ComponentFlags, ComponentId, Components, StorageType},
+    entity::{Entity, EntityLocation},
+    storage::{SparseSet, Storages},
+    world::{Mut, World},
+};
+use std::any::TypeId;
+
+pub struct EntityRef<'w> {
+    world: &'w World,
+    entity: Entity,
+    location: EntityLocation,
+}
+
+impl<'w> EntityRef<'w> {
+    #[inline]
+    pub(crate) fn new(world: &'w World, entity: Entity, location: EntityLocation) -> Self {
+        Self {
+            world,
+            entity,
+            location,
+        }
+    }
+
+    #[inline]
+    pub fn id(&self) -> Entity {
+        self.entity
+    }
+
+    #[inline]
+    pub fn location(&self) -> EntityLocation {
+        self.location
+    }
+
+    #[inline]
+    pub fn archetype(&self) -> &Archetype {
+        // SAFE: EntityRefs always point to valid entities. Valid entities always have valid archetypes
+        unsafe {
+            self.world
+                .archetypes
+                .get_unchecked(self.location.archetype_id)
+        }
+    }
+
+    #[inline]
+    pub fn contains<T: Component>(&self) -> bool {
+        self.contains_type_id(TypeId::of::<T>())
+    }
+
+    #[inline]
+    pub fn contains_id(&self, component_id: ComponentId) -> bool {
+        // SAFE: entity location is valid
+        unsafe { contains_component_with_id(self.world, component_id, self.location) }
+    }
+
+    #[inline]
+    pub fn contains_type_id(&self, type_id: TypeId) -> bool {
+        // SAFE: entity location is valid
+        unsafe { contains_component_with_type(self.world, type_id, self.location) }
+    }
+
+    #[inline]
+    pub fn get<T: Component>(&self) -> Option<&'w T> {
+        // SAFE: entity location is valid and returned component is of type T
+        unsafe {
+            get_component_with_type(self.world, TypeId::of::<T>(), self.entity, self.location)
+                .map(|value| &*value.cast::<T>())
+        }
+    }
+
+    /// # Safety
+    /// this cannot be used to produce aliased mutable access to an entity's component
```

Comment:
I think I understood this comment backwards, which can be dangerous. The safety comment of `Query::get_component_unchecked_mut()` is clearer:
```rust
    /// # Safety
    /// This allows aliased mutability. You must make sure this call does not result in multiple mutable references to the same component
```